File: modules/xml/test/model/XmlCorrectorTest.java
package model;

import static org.hamcrest.CoreMatchers.equalTo;
import static org.junit.Assert.assertNotNull;
import static org.junit.Assert.assertThat;
import static org.junit.Assert.assertTrue;

import java.io.IOException;
import java.nio.file.Files;
import java.nio.file.Path;
import java.nio.file.Paths;
import java.util.List;

import org.junit.Test;

public class XmlCorrectorTest {

  private static final Path BASE = Paths.get("test", "resources");

  private static void assertError(XmlError error, int expectedLine, XmlErrorType expectedType, String expectedMessage) {
    assertThat("Fehler sollte in Zeile " + expectedLine + " sein, war aber in Zeile " + error.getLine(),
        error.getLine(), equalTo(expectedLine));

    assertThat("Typ des Fehlers sollte " + expectedType + " sein, war aber " + error.getErrorType(),
        error.getErrorType(), equalTo(expectedType));

    assertThat(error.getErrorMessage(), equalTo(expectedMessage));
  }

  private static void assertErrorNum(int errorNum, int expectedErrors) {
    assertThat("Es sollte(n) " + expectedErrors + " Fehler sein, es sind aber " + errorNum + "!", errorNum,
        equalTo(expectedErrors));
  }

  private static String readFile(Path file) {
    try {
      return String.join("\n", Files.readAllLines(file));
    } catch (IOException e) {
      return "";
    }
  }

  @Test
  public void emptyXmlAgainstXSD() {
    Path xsd = Paths.get(BASE.toString(), "note.xsd");

    List<XmlError> out = XmlCorrector.correct("", readFile(xsd), XmlExType.XML_XSD);

    assertErrorNum(out.size(), 1);
    assertError(out.get(0), -1, XmlErrorType.FATALERROR, "Premature end of file.");
  }

  @Test
  public void testClosingTagMissingXmlDtd() {
    Path xml = Paths.get(BASE.toString(), "partyNoClosingTag.xml");

    List<XmlError> out = XmlCorrector.correct(readFile(xml), "", XmlExType.XML_DTD);

    assertNotNull(out);
    assertErrorNum(out.size(), 1);
    assertError(out.get(0), 17, XmlErrorType.FATALERROR,
        "The element type \"getraenk\" must be terminated by the matching end-tag \"</getraenk>\".");
  }

  @Test
  public void testClosingTagMissingXmlXsd() {
    Path xml = Paths.get(BASE.toString(), "noteNoClosingTag.xml");
    Path xsd = Paths.get(BASE.toString(), "note.xsd");

    List<XmlError> out = XmlCorrector.correct(readFile(xml), readFile(xsd), XmlExType.XML_XSD);

    assertNotNull(out);
    assertErrorNum(out.size(), 1);
    assertError(out.get(0), 7, XmlErrorType.FATALERROR,
        "The element type \"to\" must be terminated by the matching end-tag \"</to>\".");
  }

  @Test
  public void testCorrectDTDAgainstXML() {
    Path referenceFile = Paths.get(BASE.toString(), "party.xml");
    Path grammar = Paths.get(BASE.toString(), "party.dtd");

    List<XmlError> out = XmlCorrector.correct(readFile(referenceFile), readFile(grammar), XmlExType.XML_DTD);

    assertNotNull(out);
    assertTrue(out.isEmpty());
  }

  @Test
  public void testCorrectXMLAgainstDTD() {
    Path file = Paths.get(BASE.toString(), "party.xml");
    Path grammar = Paths.get(BASE.toString(), "party.dtd");

    List<XmlError> out = XmlCorrector.correct(readFile(file), readFile(grammar), XmlExType.XML_DTD);

    assertNotNull(out);
    assertTrue(out.isEmpty());
  }

  @Test
  public void testCorrectXMLAgainstXSD() {
    Path xml = Paths.get(BASE.toString(), "note.xml");
    Path xsd = Paths.get(BASE.toString(), "note.xsd");

    List<XmlError> out = XmlCorrector.correct(readFile(xml), readFile(xsd), XmlExType.XML_XSD);

    assertNotNull(out);
    assertTrue(out.isEmpty());
  }

  @Test
  public void testMissingAttributeXmlDtd() {
    // Path file = Paths.get(BASE.toString(), "partyMissingAttribute.xml");
    //
    // List<XmlError> out = XmlCorrector.correct(readFile(file), "",
    // XmlExType.XML_DTD);
    //
    // assertErrorNum(out.size(), 1);
    // assertError(out.get(0), 7, XmlErrorType.ERROR,
    // "Attribute \"name\" is required and must be specified for element type
    // \"gast\".");
  }

  @Test
  public void testMissingTagXmlXsd() {
    Path xml = Paths.get(BASE.toString(), "noteMissingTag.xml");
    Path xsd = Paths.get(BASE.toString(), "note.xsd");

    List<XmlError> out = XmlCorrector.correct(readFile(xml), readFile(xsd), XmlExType.XML_XSD);

    assertErrorNum(out.size(), 1);
    assertError(out.get(0), 5, XmlErrorType.ERROR,
        "cvc-complex-type.2.4.a: Invalid content was found starting with element 'body'. One of '{heading}' is expected.");
  }

  @Test
  public void testNoRootXmlDtd() {
    Path file = Paths.get(BASE.toString(), "partyNoRoot.xml");

    List<XmlError> out = XmlCorrector.correct(readFile(file), "", XmlExType.XML_DTD);

    assertErrorNum(out.size(), 1);
    assertError(out.get(0), 17, XmlErrorType.FATALERROR,
        "The markup in the document following the root element must be well-formed.");
  }

  @Test
  public void testNoRootXmlXsd() {
    Path xml = Paths.get(BASE.toString(), "noteNoRoot.xml");
    Path xsd = Paths.get(BASE.toString(), "note.xsd");

    List<XmlError> out = XmlCorrector.correct(readFile(xml), readFile(xsd), XmlExType.XML_XSD);

    assertErrorNum(out.size(), 2);
    assertError(out.get(0), 2, XmlErrorType.ERROR, "cvc-elt.1.a: Cannot find the declaration of element 'to'.");
    assertError(out.get(1), 3, XmlErrorType.FATALERROR,
        "The markup in the document following the root element must be well-formed.");
  }

  @Test
  public void testWrongAttributeDtdXml() {
    // Path referenceFile = Paths.get(BASE.toString(), "partyNoDate.xml");
    //
    // List<XmlError> out = XmlCorrector.correct(readFile(referenceFile), "",
    // XmlExType.DTD_XML);
    //
    // assertErrorNum(out.size(), 1);
    // assertError(out.get(0), 10, XmlErrorType.ERROR, "Attribute \"datum\" must
    // be declared for element type \"party\".");
  }

  @Test
  public void testWrongTagXmlDtd() {
    // Path file = Paths.get(BASE.toString(), "partyWrongTag.xml");
    //
    // List<XmlError> out = XmlCorrector.correct(readFile(file), "",
    // XmlExType.XML_DTD);
    //
    // assertErrorNum(out.size(), 2);
    // assertError(out.get(0), 2, XmlErrorType.ERROR, "Element type \"guest\"
    // must be declared.");
    // assertError(out.get(1), 13, XmlErrorType.ERROR, "The content of element
    // type \"party\" must match \"(gast)*\".");
  }

  @Test
  public void testWrongTagXmlXsd() {
    Path xml = Paths.get(BASE.toString(), "noteWrongTag.xml");
    Path xsd = Paths.get(BASE.toString(), "note.xsd");

    List<XmlError> out = XmlCorrector.correct(readFile(xml), readFile(xsd), XmlExType.XML_XSD);

    assertErrorNum(out.size(), 1);
    assertError(out.get(0), 11, XmlErrorType.ERROR,
        "cvc-complex-type.2.4.a: Invalid content was found starting with element 'sender'. One of '{from}' is expected.");
  }

  @Test
  public void xmlNoElement() {
    String xml = readFile(Paths.get(BASE.toString(), "xmlNoElement.xml"));

    List<XmlError> out = XmlCorrector.correct(xml, "", XmlExType.XML_DTD);

    assertErrorNum(out.size(), 1);
    assertError(out.get(0), -1, XmlErrorType.FATALERROR, "Premature end of file.");
  }
}

File: modules/xml/test/model/XmlCorrectorTest.scala
package model

import java.nio.file.{Files, Paths}

import org.junit.Assert.{assertNotNull, assertTrue}
import org.junit.Test
import org.scalatest.Matchers._

import scala.collection.JavaConverters._
import scala.util.Try

class XmlCorrectorTest {

  private val BASE = Paths.get("test", "resources")

  private def assertError(error: XmlError, expectedLine: Int, expectedMessage: String): Unit = {
    error.line shouldBe expectedLine

    error.errorMessage shouldBe expectedMessage
  }

  private def assertErrorNum(errorNum: Int, expectedErrors: Int): Unit = errorNum shouldBe expectedErrors


  private def readFile(file: java.nio.file.Path): String = Try(Files.readAllLines(file).asScala.mkString("\n")).getOrElse("")

  @Test
  def emptyXmlAgainstXSD() {
    val xsd = Paths.get(BASE.toString, "note.xsd")

    val out: List[XmlError] = XmlCorrector.correct("", readFile(xsd), XML_XSD)

    assertErrorNum(out.size, 1)
    assertError(out.head, -1, "Premature end of file.")
  }

  @Test
  def testClosingTagMissingXmlDtd() {
    val xml = Paths.get(BASE.toString, "partyNoClosingTag.xml")

    val out: List[XmlError] = XmlCorrector.correct(readFile(xml), "", XML_DTD)

    assertNotNull(out)
    assertErrorNum(out.size, 1)
    assertError(out.head, 17,
      "The element type \"getraenk\" must be terminated by the matching end-tag \"</getraenk>\".")
  }

  @Test
  def testClosingTagMissingXmlXsd() {
    val xml = Paths.get(BASE.toString, "noteNoClosingTag.xml")
    val xsd = Paths.get(BASE.toString, "note.xsd")

    val out: List[XmlError] = XmlCorrector.correct(readFile(xml), readFile(xsd), XML_XSD)

    assertNotNull(out)
    assertErrorNum(out.size, 1)
    assertError(out.head, 7, "The element type \"to\" must be terminated by the matching end-tag \"</to>\".")
  }

  @Test
  def testCorrectDTDAgainstXML() {
    val referenceFile = Paths.get(BASE.toString, "party.xml")
    val grammar = Paths.get(BASE.toString, "party.dtd")

    val out: List[XmlError] = XmlCorrector.correct(readFile(referenceFile), readFile(grammar), XML_DTD)

    assertNotNull(out)
    assertTrue(out.isEmpty)
  }

  @Test
  def testCorrectXMLAgainstDTD() {
    val file = Paths.get(BASE.toString, "party.xml")
    val grammar = Paths.get(BASE.toString, "party.dtd")

    val out: List[XmlError] = XmlCorrector.correct(readFile(file), readFile(grammar), XML_DTD)

    assertNotNull(out)
    assertTrue(out.isEmpty)
  }

  @Test
  def testCorrectXMLAgainstXSD() {
    val xml = Paths.get(BASE.toString, "note.xml")
    val xsd = Paths.get(BASE.toString, "note.xsd")

    val out: List[XmlError] = XmlCorrector.correct(readFile(xml), readFile(xsd), XML_XSD)

    assertNotNull(out)
    assertTrue(out.isEmpty)
  }

  @Test
  def testMissingAttributeXmlDtd() {
    // val file = Paths.get(BASE.toString, "partyMissingAttribute.xml")
    //
    // List<XmlError> out = XmlCorrector.correct(readFile(file), "",
    // XmlExType.XML_DTD)
    //
    // assertErrorNum(out.size, 1)
    // assertError(out.head, 7, XmlErrorType.ERROR,
    // "Attribute \"name\" is required and must be specified for element type
    // \"gast\".")
  }

  @Test
  def testMissingTagXmlXsd() {
    val xml = Paths.get(BASE.toString, "noteMissingTag.xml")
    val xsd = Paths.get(BASE.toString, "note.xsd")

    println(xml.toFile.exists + " :: " + xsd.toFile.exists)

    val out: List[XmlError] = XmlCorrector.correct(readFile(xml), readFile(xsd), XML_XSD)

    assertErrorNum(out.size, 1)
    assertError(out.head, 5,
      "cvc-complex-type.2.4.a: Invalid content was found starting with element 'body'. One of '{heading}' is expected.")
  }

  @Test
  def testNoRootXmlDtd() {
    val file = Paths.get(BASE.toString, "partyNoRoot.xml")

    val out: List[XmlError] = XmlCorrector.correct(readFile(file), "", XML_DTD)

    assertErrorNum(out.size, 1)
    assertError(out.head, 17, "The markup in the document following the root element must be well-formed.")
  }

  @Test
  def testNoRootXmlXsd() {
    val xml = Paths.get(BASE.toString, "noteNoRoot.xml")
    val xsd = Paths.get(BASE.toString, "note.xsd")

    val out: List[XmlError] = XmlCorrector.correct(readFile(xml), readFile(xsd), XML_XSD)

    assertErrorNum(out.size, 2)
    assertError(out.head, 2, "cvc-elt.1.a: Cannot find the declaration of element 'to'.")
    assertError(out(1), 3, "The markup in the document following the root element must be well-formed.")
  }

  @Test
  def testWrongAttributeDtdXml() {
    // val referenceFile = Paths.get(BASE.toString, "partyNoDate.xml")
    //
    // List<XmlError> out = XmlCorrector.correct(readFile(referenceFile),
    // "",
    // XmlExType.DTD_XML)
    //
    // assertErrorNum(out.size, 1)
    // assertError(out.head, 10, XmlErrorType.ERROR, "Attribute \"datum\" must
    // be declared for element type \"party\".")
  }

  @Test
  def testWrongTagXmlDtd() {
    // val file = Paths.get(BASE.toString, "partyWrongTag.xml")
    //
    // List<XmlError> out = XmlCorrector.correct(readFile(file), "",
    // XmlExType.XML_DTD)
    //
    // assertErrorNum(out.size, 2)
    // assertError(out.head, 2, XmlErrorType.ERROR, "Element type \"guest\"
    // must be declared.")
    // assertError(out(1), 13, XmlErrorType.ERROR, "The content of element
    // type \"party\" must match \"(gast)*\".")
  }

  @Test
  def testWrongTagXmlXsd() {
    val xml = Paths.get(BASE.toString, "noteWrongTag.xml")
    val xsd = Paths.get(BASE.toString, "note.xsd")

    val out: List[XmlError] = XmlCorrector.correct(readFile(xml), readFile(xsd), XML_XSD)

    assertErrorNum(out.size, 1)
    assertError(out.head, 11,
      "cvc-complex-type.2.4.a: Invalid content was found starting with element 'sender'. One of '{from}' is expected.")
  }

  @Test
  def testXmlNoElement() {
    val xml = readFile(Paths.get(BASE.toString, "xmlNoElement.xml"))

    val out: List[XmlError] = XmlCorrector.correct(xml, "", XML_DTD)

    assertErrorNum(out.size, 1)
    assertError(out.head, -1, "Premature end of file.")
  }
}
